lib/xdrgen/generators/go.rb

```diff
@@ -549,6 +531,256 @@ def render_encode_to_body(out, var, type, self_encode:)
         out.puts "  }"
       end
 
+      def render_struct_decode_from_interface(out, struct)
+        name = name(struct)
+        out.puts "// DecodeFrom decodes this value using the Decoder."
+        out.puts "func (s *#{name}) DecodeFrom(d *xdr.Decoder) (int, error) {"
+        out.puts "  var err error"
+        out.puts "  var n, nTmp int"
+        declared_variables = []
+        struct.members.each do |m|
+          mn = name(m)
+          render_decode_from_body(out, "s.#{mn}", m.type, declared_variables: declared_variables, self_encode: false)
+        end
+        out.puts "  return n, nil"
+        out.puts "}"
+        out.break
+      end
+
+      def render_union_decode_from_interface(out, union)
+        name = name(union)
+        out.puts "// DecodeFrom decodes this value using the Decoder."
+        out.puts "func (u *#{name}) DecodeFrom(d *xdr.Decoder) (int, error) {"
+        out.puts "  disc, n, err := d.DecodeInt()"
+        out.puts "  if err != nil {"
+        out.puts "    return 0, err"
+        out.puts "  }"
+        out.puts "  u.#{name(union.discriminant)} = #{reference union.discriminant.type}(disc)"
```

Comment:
I mentioned this on Slack but you should check here if `disc` is actually a valid value for this union. Something like what `go-xdr` is doing here: https://github.com/stellar/go-xdr/blob/8017fc4bdfee412460b17f99d7d56fde3ab29047/xdr3/decode.go#L541-L555 Without it malicious actor could pass invalid value crashing a decoder.

Reply:
We check for it, see the generated code.

Reply:
This does the right thing according to the spec. The type has a default void value, which means that the type is considered void if the discriminant is different to `CreateAccountResultCodeCreateAccountSuccess`. You can check the XDR spec.

Reply:
The definition of the type is:

```
// CreateAccountResult is an XDR Union defines as:
//
//   union CreateAccountResult switch (CreateAccountResultCode code)
//    {
//    case CREATE_ACCOUNT_SUCCESS:
//        void;
//    default:
//        void;
//    };
//
```

So, as unintuitive as this may seem, you can use whichever value we want other than `CREATE_ACCOUNT_SUCCESS` and it will still be considered valid.

Reply:
Note that for unions without the default arm, we also do the right thing thanks to the error returned at the end.

Reply:
I think that using default arms is questionable. We can chose to change the XDR definitions to disable default arms, but that's a totally different issue.

Reply:
I disagree this is a valid code according to xdr specification. See https://datatracker.ietf.org/doc/html/rfc4506.html#section-4.15:
```
Discriminated unions
   are declared as follows:

         union switch (discriminant-declaration) {
         case discriminant-value-A:
            arm-declaration-A;
         case discriminant-value-B:
            arm-declaration-B;
         ...
         default: default-declaration;
         } identifier;

   Each "case" keyword is followed by a legal value of the discriminant.
   The default arm is optional. If it is not specified, then a valid
   encoding of the union cannot take on unspecified discriminant values.

```
Each case needs to be followed by a _legal_ value of the discriminant and it cannot take on unspecified discriminant values.

Reply:
Note how `ArmForSwitch()` (which is what `go-xdr` uses for union arm checking) does the exact same thing:

```go
func (u CreateAccountResult) ArmForSwitch(sw int32) (string, bool) {
	switch CreateAccountResultCode(sw) {
	case CreateAccountResultCodeCreateAccountSuccess:
		return "", true
	default:
		return "", true
	}
}
```

Reply:
`ArmForSwitch` is not used in the code I linked here: https://github.com/stellar/go-xdr/blob/8017fc4bdfee412460b17f99d7d56fde3ab29047/xdr3/decode.go#L541-L555. I think spec is pretty clear on the matter we discuss here and the old code returns error on invalid discriminant values.

Reply:
Here is another example without default value:

```
// Memo is an XDR Union defines as:
//
//   union Memo switch (MemoType type)
//    {
//    case MEMO_NONE:
//        void;
//    case MEMO_TEXT:
//        string text<28>;
//    case MEMO_ID:
//        uint64 id;
//    case MEMO_HASH:
//        Hash hash; // the hash of what to pull from the content server
//    case MEMO_RETURN:
//        Hash retHash; // the hash of the tx you are rejecting
//    };
//
```

Note how both `ArmForSwitch` do return an error in this case:

```go
// ArmForSwitch returns which field name should be used for storing
// the value for an instance of Memo
func (u Memo) ArmForSwitch(sw int32) (string, bool) {
	switch MemoType(sw) {
	case MemoTypeMemoNone:
		return "", true
	case MemoTypeMemoText:
		return "Text", true
	case MemoTypeMemoId:
		return "Id", true
	case MemoTypeMemoHash:
		return "Hash", true
	case MemoTypeMemoReturn:
		return "RetHash", true
	}
	return "-", false
}
```

`go-xdr` will interpret "-" as an error at https://github.com/stellar/go-xdr/blob/master/xdr3/decode.go#L577-L586 

```
// DecodeFrom decodes this value using the Decoder.
func (u *Memo) DecodeFrom(d *xdr.Decoder) (int, error) {
	disc, n, err := d.DecodeInt()
	if err != nil {
		return 0, err
	}
	u.Type = MemoType(disc)
	switch MemoType(u.Type) {
	case MemoTypeMemoNone:
		// Void
		return n, nil
	case MemoTypeMemoText:
		u.Text = new(string)
		var nTmp int
		(*u.Text), nTmp, err = d.DecodeString(28)
		n += nTmp
		if err != nil {
			return n, err
		}
		return n, nil
	case MemoTypeMemoId:
		u.Id = new(Uint64)
		var nTmp int
		nTmp, err = (*u.Id).DecodeFrom(d)
		n += nTmp
		if err != nil {
			return n, err
		}
		return n, nil
	case MemoTypeMemoHash:
		u.Hash = new(Hash)
		var nTmp int
		nTmp, err = (*u.Hash).DecodeFrom(d)
		n += nTmp
		if err != nil {
			return n, err
		}
		return n, nil
	case MemoTypeMemoReturn:
		u.RetHash = new(Hash)
		var nTmp int
		nTmp, err = (*u.RetHash).DecodeFrom(d)
		n += nTmp
		if err != nil {
			return n, err
		}
		return n, nil
	}
	return n, fmt.Errorf("Type (MemoType) switch value '%d' is not valid for union Memo", u.Type)
}
```

Reply:
A few lines above the code you linked (that I linked directly in 2 comments above) there's a check using `ValidEnum` method that ensures that discriminant value is valid and I think we need the same thing in the new code. This is required by XDR spec and without this check invalid XDR can be decoded without errors which should not happen.

Reply:
@bartekn I just updated the code to remove redundant return values when there is a default branch (In case you notice the code changed)

Reply:
OK, I read entire thread again. It seems that the old decoder's code is wrong then (because it doesn't allow any value on `default` case). Sorry for the noise!

Reply:
Ah, good point! I will incorporate that to the code.

I also noticed the `DecodeEnum` function which validates enums, I will incorporate that too.

Reply:
OK, this should be fixed, PTAL